cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+	err := h.RunScripts(ctx, HookTypePre, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	err = actionFn()
+	if err != nil {
+		return err
+	}
+
+	err = h.RunScripts(ctx, HookTypePost, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	return nil
+}
+
+// Invokes any registered script hooks for the specified hook type and command.
+func (h *CommandHooks) RunScripts(ctx context.Context, hookType HookType, commands []string) error {
+	scripts := h.getScriptsForHook(hookType, commands)
+	for _, scriptConfig := range scripts {
+		err := h.execScript(ctx, scriptConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Gets the script to execute based on the script configuration values
+// For inline scripts this will also create a temporary script file to execute
+func (h *CommandHooks) GetScript(scriptConfig *ScriptConfig) (tools.Script, error) {
+	if scriptConfig.Location == "" {
+		if scriptConfig.Path != "" {
+			scriptConfig.Location = ScriptLocationPath
+		} else if scriptConfig.Script != "" {
+			scriptConfig.Location = ScriptLocationInline
+		}
+	}
+
+	if scriptConfig.Location == ScriptLocationInline {
+		tempScript, err := createTempScript(scriptConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		scriptConfig.Path = tempScript
+	}
+
+	if scriptConfig.Type == "" {
+		scriptType, err := inferScriptTypeFromFilePath(scriptConfig.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		scriptConfig.Type = scriptType
+	}
+
+	switch scriptConfig.Type {
+	case ScriptTypeBash:
+		return bash.NewBashScript(h.commandRunner, h.cwd, h.envVars), nil
+	case ScriptTypePowershell:
+		return powershell.NewPowershellScript(h.commandRunner, h.cwd, h.envVars), nil
+	default:
+		return nil, fmt.Errorf(
+			"script type '%s' is not a valid option. Only Bash and powershell scripts are supported",
+			scriptConfig.Type,
+		)
+	}
+}
+
+func (h *CommandHooks) getScriptsForHook(prefix HookType, commands []string) []*ScriptConfig {
+	validHookNames := []string{}
+
+	for _, commandName := range commands {
+		// Convert things like `azd config list` => 'configlist`
+		commandName = strings.TrimPrefix(commandName, "azd")
+		commandName = strings.TrimSpace(commandName)
+		commandName = strings.ReplaceAll(commandName, " ", "")
+
+		validHookNames = append(validHookNames, strings.ToLower(string(prefix)+commandName))
+	}
+
+	allHooks := []*ScriptConfig{}
+	explicitHooks := h.getExplicitHooks(prefix, validHookNames)
+	implicitHooks := h.getImplicitHooks(prefix, validHookNames)
+
+	allHooks = append(allHooks, explicitHooks...)
+
+	// Only append implicit hooks that were not already wired up explicitly.
+	for _, implicitHook := range implicitHooks {
+		index := slices.IndexFunc(allHooks, func(hook *ScriptConfig) bool {
+			return implicitHook.Name == hook.Name
+		})
+
+		if index >= 0 {
+			log.Printf(
```

Comment:
This feels like an error to me, and it will help save users from authoring errors.

Reply:
We can throw an error if the same hook was implicitly and explicitly added.  The only use case that wouldn't be supported then would be the user would not be able to store their custom scripts in the `.azure/hook` folder and reference them in azure.yaml.  They would need to chose a more public location in their repo.

I don't have a strong feeling one way or the other.

Reply:
That's a good point. What if we checked if the `Path` was the same?

So, if a user conventionally creates a script in `.azure/hooks`, and decided to convert to explicit configuration (perhaps because it's more configurable), it does work. But if the `Path` was different, then it is an error (shadowing occurred).

Reply:
Removing implicit hooks for the time being.